File: apps/ml-server/src/services/redis_service.py
import redis
from redis.exceptions import TimeoutError as RedisTimeoutError
import logging
from typing import Optional
import os

logger = logging.getLogger(__name__)


class RedisClient:
    _instance: Optional["RedisClient"] = None

    def __init__(self):
        host = os.getenv("REDIS_HOST", "localhost")
        port = int(os.getenv("REDIS_PORT", "6379"))
        password = os.getenv("REDIS_PASSWORD") or None
        self.redis = redis.Redis(
            host=host,
            port=port,
            password=password,
            decode_responses=True,
            socket_connect_timeout=5,
            socket_timeout=5,
            retry_on_timeout=True,
        )
        self._is_connected = False
        self._connect()

    def _connect(self):
        try:
            self.redis.ping()
            self._is_connected = True
            logger.info("Redis client connected and ready")
        except redis.RedisError as e:
            self._is_connected = False
            logger.error("Redis connection failed", exc_info=e)

    @classmethod
    def get_instance(cls) -> "RedisClient":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def is_ready(self) -> bool:
        return self._is_connected

    def get(self, key: str) -> Optional[str]:
        return self.redis.get(key)

    def set(self, key: str, value: str, ttl_seconds: int):
        self.redis.setex(key, ttl_seconds, value)

    def set_session(self, session_id: str, value: str):
        self.set(session_id, value, 20 * 60)

    def delete(self, key: str):
        self.redis.delete(key)

    def xadd_event(
        self,
        stream_key: str,
        event_type: str,
        payload: dict,
        max_len: Optional[int] = 10000,
    ) -> Optional[str]:
        import json
        import time
        try:
            fields = {
                "event": event_type,
                "payload": json.dumps(payload, default=str),
                "ts": str(time.time()),
            }
            kwargs = {}
            if max_len:
                kwargs["maxlen"] = max_len
                kwargs["approximate"] = True
            msg_id = self.redis.xadd(stream_key, fields, **kwargs)
            return msg_id
        except redis.RedisError as e:
            logger.error("Redis XADD event failed", exc_info=e)
            return None

    def create_consumer_group(
        self, stream_key: str, group_name: str, start_id: str = "0"
    ) -> bool:
        try:
            self.redis.xgroup_create(
                stream_key, group_name, id=start_id, mkstream=True
            )
            logger.info("Consumer group %s created for stream %s", group_name, stream_key)
            return True
        except redis.RedisError as e:
            if "BUSYGROUP" in str(e):
                logger.debug("Consumer group %s already exists", group_name)
                return True
            logger.error("Redis XGROUP CREATE failed", exc_info=e)
            return False

    def read_from_group(
        self,
        stream_key: str,
        group_name: str,
        consumer_name: str,
        count: int = 1,
        block_ms: int = 5000,
    ):
        try:
            result = self.redis.xreadgroup(
                groupname=group_name,
                consumername=consumer_name,
                streams={stream_key: ">"},
                count=count,
                block=block_ms,
            )
            if not result:
                return []
            messages = result[0][1] if result else []
            return [(msg_id, dict(fields)) for msg_id, fields in messages]
        except RedisTimeoutError:
            # Block timeout with no messages - normal, not an error
            return []
        except redis.RedisError as e:
            logger.error("Redis XREADGROUP failed", exc_info=e)
            return []

    def acknowledge(self, stream_key: str, group_name: str, message_id: str) -> bool:
        try:
            self.redis.xack(stream_key, group_name, message_id)
            return True
        except redis.RedisError as e:
            logger.error("Redis XACK failed for %s", message_id, exc_info=e)
            return False

    def health_check(self) -> bool:
        try:
            return self.redis.ping() is True
        except redis.RedisError as e:
            logger.error("Redis health check failed", exc_info=e)
            return False

    def close(self):
        try:
            self.redis.close()
            logger.info("Redis connection closed gracefully")
        except redis.RedisError as e:
            logger.error("Error closing Redis connection", exc_info=e)

File: apps/api/src/services/redis.service.ts
import { Redis } from 'ioredis';
import logger from '../core/logger';
import { redisConfig } from '../config';

export const AI_QUEUE_CONSUMER_GROUP = 'ai-workers';
export const AI_QUEUE_CONSUMER_NAME = 'worker-1';


class RedisClient {
    private static instance: RedisClient;
    private redis: Redis;
    private isConnected = false;

    private constructor() {
        this.redis = new Redis({
            host: redisConfig.redisHost,
            port: redisConfig.redisPort,
            password: redisConfig.redisPassword,
        });
        this.setupEventHandlers();
    }

    private setupEventHandlers(): void {
        if (!this.redis) return;

        this.redis.on('connect', () => {
            this.isConnected = true;
            logger.info('Redis client connected');
        });

        this.redis.on('ready', () => {
            logger.info('Redis client ready');
        });

        this.redis.on('error', (error: Error) => {
            logger.error('Redis client error', {
                error: error.message,
                stack: error.stack,
            });
        });

        this.redis.on('close', () => {
            this.isConnected = false;
            logger.warn('Redis client connection closed');
        });

        this.redis.on('reconnecting', () => {
            logger.info('Redis client reconnecting...');
        });

        this.redis.on('end', () => {
            this.isConnected = false;
            logger.warn('Redis client connection ended');
        });
    }

    public static getInstance() {
        if (!this.instance) {
            this.instance = new RedisClient();
        }
        return this.instance;
    }

    public isReady() {
        return this.isConnected && this.redis.status === 'ready';
    }

    public async getResult(key: string) {
        return await this.redis.get(key);
    }

    // 20 minutes
    public async setSession(sessionId: string, value: string) {
        await this.redis.set(sessionId, value, 'EX', 20 * 60);
    }

    /**
     * Redis Health check
     */
    public async healthCheck(): Promise<boolean> {
        try {
            if (!this.isReady()) {
                return false;
            }

            const result = await this.redis.ping();
            return result === 'PONG';
        } catch (error) {
            const message =
                error instanceof Error ? error.message : 'Unknown error';
            logger.error('Redis health check failed', {
                error: message,
            });
            return false;
        }
    }

    public async deleteKey(key: string) {
        await this.redis.del(key);
    }

    /**
     * Add message to a stream (for AI job queue producer).
     * @param streamKey - Stream name
     * @param fields - Field key-value pairs (values will be stringified if not strings)
     * @param maxLen - Optional max stream length (approximate)
     */
    public async addToStream(
        streamKey: string,
        fields: Record<string, string | number | object>,
        maxLen?: number,
    ): Promise<string | null> {
        try {
            const flat: string[] = [];
            for (const [k, v] of Object.entries(fields)) {
                flat.push(k, typeof v === 'object' ? JSON.stringify(v) : String(v));
            }
            const args: string[] = maxLen
                ? ['MAXLEN', '~', String(maxLen), '*', ...flat]
                : ['*', ...flat];
            const id = await this.redis.xadd(streamKey, ...args);
            return id as string | null;
        } catch (error) {
            const message =
                error instanceof Error ? error.message : 'Unknown error';
            logger.error('Redis XADD failed', { error: message, streamKey });
            return null;
        }
    }

    /**
     * Create consumer group if it doesn't exist
     */
    public async createConsumerGroup(
        streamKey: string,
        groupName: string,
        startId: string = '0',
    ): Promise<void> {
        try {
            await this.redis.xgroup(
                'CREATE',
                streamKey,
                groupName,
                startId,
                'MKSTREAM',
            );
            logger.info(
                `Consumer group "${groupName}" created for stream "${streamKey}"`,
            );
        } catch (error) {
            // BUSYGROUP error means group already exists, which is fine
            if (error instanceof Error && error.message.includes('BUSYGROUP')) {
                logger.info(`Consumer group "${groupName}" already exists`);
            } else {
                throw error;
            }
        }
    }

    /**
     * Read messages from stream using consumer group
     */
    public async readFromGroup(
        streamKey: string,
        groupName: string,
        consumerName: string,
        count: number = 10,
        blockMs: number = 5000,
    ) {
        try {
            const results = await this.redis.xreadgroup(
                'GROUP',
                groupName,
                consumerName,
                'COUNT',
                count,
                'BLOCK',
                blockMs,
                'STREAMS',
                streamKey,
                '>', // Read new messages not yet delivered to this consumer group
            );

            return results;
        } catch (error) {
            logger.error('Error reading from consumer group', {
                error: error instanceof Error ? error.message : 'Unknown error',
                streamKey,
                groupName,
                consumerName,
            });
            throw error;
        }
    }

    /**
     * Acknowledge message processing
     */
    public async acknowledgeMessage(
        streamKey: string,
        groupName: string,
        messageId: string,
    ): Promise<void> {
        await this.redis.xack(streamKey, groupName, messageId);
    }

    /**
     * Get pending messages for this consumer
     */
    public async getPendingMessages(
        streamKey: string,
        groupName: string,
        consumerName: string,
        count: number = 10,
    ) {
        try {
            // Get pending message IDs
            const pending = await this.redis.xpending(
                streamKey,
                groupName,
                '-',
                '+',
                count,
                consumerName,
            );

            if (!pending || pending.length === 0) {
                return null;
            }

            // Claim these messages
            const messageIds = pending.map((p: never) => p[0]);
            const results = await this.redis.xclaim(
                streamKey,
                groupName,
                consumerName,
                60000, // Min idle time in ms (1 minute)
                ...messageIds,
            );

            return results;
        } catch (error) {
            logger.error('Error getting pending messages', {
                error: error instanceof Error ? error.message : 'Unknown error',
            });
            return null;
        }
    }

    /**
     * Close Redis connection
     */
    public async close(): Promise<void> {
        try {
            await this.redis.quit();
            logger.info('Redis client connection closed gracefully');
        } catch (error) {
            const message =
                error instanceof Error ? error.message : 'Unknown error';
            logger.error('Error closing Redis connection', {
                error: message,
            });
            // Force close if graceful close fails
            this.redis.disconnect();
        }
    }

    /**
     * Get underlying Redis client for advanced operations
     */
    public getClient(): Redis {
        return this.redis;
    }
}

export const redisClient = RedisClient.getInstance();